### DISCUSSION TITLE
讨论:  新功能“桌面视频动态背景”的实现方案

### DISCUSSION THREAD

**Discussion by jiangzc** [linuxdeepin/developer-center#2962] [Features Request & Ideas | 特性请求 & 头脑风暴]
社区里的一些朋友希望在桌面上设置动态背景，类似于 Windows上的WallpaperEngine，在桌面图标后面播放视频。

我把dde-desktop相关部分的源代码看了一下，设计出了一个巧妙的方案。

修改 dde-desktop/view/background* 模块可以创造一个透明背景，再把播放器窗口设置为桌面，桌面图标位于视频的上层。

具体的做法是：dde-desktop 启动时如果开启了 --video-wallpaper命令行选项，则选择一个特定的外部窗口(播放器)作为桌面，backgroundWidget设置透明背景、Frameless、StayOnBottom。

修改后的dde-desktop带参数启动后，在其他软件配合下实现动态背景的效果。（dde-desktop本身不能播放视频）

这个实现方案相当优秀。它对原有代码修改少、与播放器的耦合程度很低、易于扩展功能、方便其他人维护。唯一的缺点是：返回桌面的时候，（播放器窗口）桌面会把桌面图标挡住。这个也容易解决，只要当播放器窗口激活后，再激活dde-desktop窗口即可。

代码在我fork的仓库中 feat-video-wallpaper 分支里。它基于v20最新版，即基于commit 15325552572d1d5c53387c54763f57c4e13bd1e9 (tag: 5.2.0.46) 。

https://github.com/jiangzc/dde-file-manager/tree/feat-video-wallpaper

**Comment by jiangzc** [comment]
测试用例：

[video_testcase.zip](https://github.com/linuxdeepin/dde-file-manager/files/5280715/video_testcase.zip)

动态壁纸桌面测试文件

使用之前检查： （系统环境 v20 1002）  
+ 安装 xdotool
+ 当前目录中的 ./dde-desktop 是否能正常运行
+ 当前目录中的 ./video 是否能正常运行


在终端中运行 a.out  过3秒后出现动态壁纸桌面。 

a.out 是 monitor.cpp 编译后生成的可执行文件，编译方法在源代码的注释中。

**Comment by jiangzc** [comment]
我修改的内容和 5.2.0.46 代码对比：

https://github.com/linuxdeepin/dde-file-manager/compare/5.2.0.46...jiangzc:feat-video-wallpaper

**Comment by jiangzc** [comment]
@xmuli @justforlxz @felixonmars @z966768

**Comment by XMuli** [comment]
@deepin-CD 这个视频桌面的接口，是否能够和产品、设计那边确认下，合适的话，可以吸收纳入到主线中？

**Comment by shenmo7192** [comment]
> @deepin-CD 这个视频桌面的接口，是否能够和产品、设计那边确认下，合适的话，可以吸收纳入到主线中？

请问结果出来了么？
https://gitee.com/moriv4/dde-desktop-video

**Comment by XMuli** [comment]
可以将联系方式通过邮件或者QQ 或 其他方式私发给我否? [左侧有我的联系方式](https://xmuli.tech/) 然后方便沟通, 我可以帮合并进入 社区版本的分支~ (但是需要略家修改, 默认只能能在 社区版本生效, 对于专业版本, uos 版本 默认处于关闭状态, 或许等待时间会比较长~  ),

**Comment by zccrs** [comment]
确认下现在是否已经有了相关功能

**Comment by shenmo7192** [comment]
> 确认下现在是否已经有了相关功能

原项目已经不再更新，现在dde的动态壁纸功能由`fantascene`软件实现。实现的方式是Kill掉系统的dde-desktop并启动已经改造好的魔改dde-desktop

项目地址为：https://gitee.com/liuminghang/fantascene-dynamic-wallpaper

目前这个项目也被UOS当做一个社区提供的动态壁纸解决方案
然而，更新dde-desktop的时候，代码就需要重新修改。如果dde-desktop原生支持使用其他程序作为背景，会大大降低开发者的工作量，也可以更好地与系统集成，可能也能解锁更多玩法

**Reply by zccrs** [reply]
还有个方案，dde-desktop允许设置自己不显示任何背景（背景透明），把动态壁纸的窗口显示到dde-desktop后面，不过这样只支持X11。

**Comment by meiyixiang** [comment]
谢谢你的反馈，这个功能我们需要评估一下如何合入

**Comment by meiyixiang** [comment]
此功能建议已经采纳但被定义为不是很高优先级的需求后期在v23版本会实现

**Reply by shenmo7192** [reply]
收到
